hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java

```diff
@@ -411,16 +412,18 @@ public long tryReserveForUpdate(K key) {
     }
 
     @Override
-    public V tryPublishReserved(K key, final V value, final long reservationId, boolean deserialize) {
-        return updateAndGetReserved(key, value, reservationId, deserialize);
+    public V tryPublishReserved(K key, Data keyData, V value, long reservationId, boolean deserialize) {
+        return updateAndGetReserved(key, keyData, value, reservationId, deserialize);
     }
 
-    protected R updateReservedRecordInternal(K key, V value, R reservedRecord, long reservationId) {
+    protected R updateReservedRecordInternal(K key, Data keyData, V value, R reservedRecord, long reservationId) {
         if (!reservedRecord.casRecordState(reservationId, UPDATE_STARTED)) {
             return reservedRecord;
         }
 
         updateRecordValue(reservedRecord, value);
+        initInvalidationMetaData(reservedRecord, key, keyData);
```

Comment:
This is the more critical change, since the invalidation metadata is now changed on the record update, not on the record creation. But as you can see the record is not accessible (aka readable) before the `reservedRecord.casRecordState(UPDATE_STARTED, READ_PERMITTED);` is called, so it should not matter that the meta data is not set directly on the creation.

I also checked that the metadata is not used before the record is in READ_PERMITTED state:
```java
    @Override
    public V get(K key) {
        checkAvailable();

        R record = null;
        V value = null;
        try {
            record = getRecord(key);
            if (record != null) {
                if (record.getRecordState() != READ_PERMITTED) {
                    return null;
                }
                if (staleReadDetector.isStaleRead(key, record)) {
                    remove(key);
                    return null;
                }
(...)
```
So it should be okay to have the metadata set in the `updateReservedRecordInternal()` method. I also think it doesn't make a difference that the metadata itself (UUID and invaldation sequence) is now taken from the `MetaDataContainer` after the remove call, but this should be confirmed by Ahmet.